docs/proposals/202005_query_logging.md

```diff
@@ -0,0 +1,165 @@
+```
+
+Here, we can see that this log was taken before the Prometheus instance died unnaturally, and the query logged under the component **ActiveQueryTracker**, logs the respective query that did the damage. Even though it does not pin-points the exact issue that created the problem, still it is quite helpful in considering this issue as one of the potential queries that led to the unnatural death of the Prometheus instance.
+
+Another requirement is using **Audit and Adaptive Logging** in Thanos. This is a useful feature, as the former helps in aggregating all the queries made until now, and that can be utilised to track down and provide a comprehensive overview of what is happening.
+
+The latter is quite useful, as we can log all queries based upon fulfilling a certain criteria/filters, and later those can be used for inspecting abnormal queries and improving the user experience overall.
+
+
+## Proposal
+
+### Query Logging in Thanos
+
+The proposal for Query Logging has been divided into three major parts. The *first of the two logging feature* would be implemented across the Store API level, while the *last one* would be local to each component of Thanos.
+
+![](../img/thanos_log_limit.png)
+
+Each of the use case of the sub-part would be described and a possible implementation would be discussed -
+
+#### Use case 1: Audit Logging
+
+**Audit logging** is a kind of logging, where we log in every internal API requests. Typically, rather than logging in specific queries, we ought to log in every query that has been made until now.
+
+From a developer’s perspective, audit logs can keep you sane by giving you some insight into how a complex system arrived at its current state. Developers are typically fond of state diagrams and deterministic outcomes. An audit log that shows that a person or process applied a change can give developers valuable insight.
+
+#### Implementation:
+
+```go
+type AuditLogging struct {
+
+     // logger object derived from the base logger class
+     logger         log.Logger
+     // component using the audit logger
+     component      component.StoreAPI
+
+}
+```
+
+* Currently, we are looking to implement Audit Logging at the **StoreAPI** level.
+* We will use a **middleware** to intercept all the queries made to the StoreAPI.
+* As StoreAPI is interconnected with each of the other components, we can leverage this to provide a Global overview of different queries which would be useful for auditing purposes.
+* We will have a *different logger* derived from the base logger, and would log the queries separately.
+
+#### Use case 2: Adaptive Logging
+
+**Adaptive logging** is totally different from the Audit Logging. Out here in adaptive logging, we would log only certain queries which *satisfies a certain policy/satisfies a certain condition*. Possible filters could be, queries making an invalid request, latency time crossing a certain barrier, and so on.
```

Comment:
What would a policy look like? Is it a condition in the code? Is it a metric? How would user define a policy? Or is it something hard-coded?

Reply:
I am not sure how would we implement the policy, maybe first we need to fix up what sort of filters are we looking to add, or do we provide the user the flexibility to add in their own filters? What do you think?

Reply:
> // The decider logic would be implemented in the grpc-middleware itself

I think this tells us something, but I think we can define what policies we want now, to know all requirements.

Reply:
Added a policy bullet